File: pyround1/563.二叉树的坡度.py
#
# @lc app=leetcode.cn id=563 lang=python
#
# [563] 二叉树的坡度
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.slope = 0

    def findTilt(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """

        ans = 0
        def f(r):
            if not r:
                return 0
            sumleft, sumright = f(r.left), f(r.right)
            nonlocal ans
            ans += abs(sumleft - sumright)
            return sumleft + sumright + r.val
        f(root)
        return ans

# @lc code=end

File: 563.二叉树的坡度.cpp
/*
 * @lc app=leetcode.cn id=563 lang=cpp
 *
 * [563] 二叉树的坡度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int total_slope = 0;
    int findTilt(TreeNode* root) {
        helper(root);
        return total_slope;
    }
    int helper(TreeNode* root){
        if(!root)
            return 0;
        if(!root->left and !root->right)
            return root->val;

        int L = helper(root->left);
        int R = helper(root->right);
        total_slope += abs(L-R);
        root->val = root->val + L + R;
        return root->val;
        
    }
};
// @lc code=end

